### ISSUE TITLE
Expert-granularity CPU offload for quantized MoE: 30B-A3B QLoRA in ~7 GB (single resident expert layer, extends the layer_offloading idea)

### ISSUE THREAD

**Issue by pjordanandrsn** [axolotl-ai-cloud/axolotl#3787]
## Use case

`quantize_moe_experts` (#3439) got fused-expert 4-bit working, and #3747's selective dequant lowered
the ceiling further — but the floor for a 30B-class MoE QLoRA is still ~17–24 GiB reserved, so
12–16 GB cards are out. Demand for exactly that exists in the tracker (#3530 OOMs a 32 GB card on
gpt-oss-20b; #3402 asks for lower-VRAM MoE training generally).

The missing piece is granularity: `layer_offloading` (#3512) already proves the pinned-CPU +
prefetch-stream pattern in-tree, but at whole-decoder-layer scope its measured savings are modest
(~3 GB on Qwen3.5-35B) because attention/router params are small — in a fine-tuned MoE, ~90 % of
frozen bytes are the experts. Offloading **only the quantized expert tensors, with a single
resident expert layer on GPU** and prefetch of layer N+1 during layer N's compute, moves almost all
of those bytes off-device while touching nothing else.

## Evidence it works (measured, single GPU)

I've implemented this out of tree ([experts4bit-qlora](https://github.com/pjordanandrsn/experts4bit-qlora),
MIT) against transformers-v5 fused experts, and measured on a 12 GB RTX A2000:

| model | peak VRAM with expert offload | without |
|---|---|---|
| Qwen3-30B-A3B QLoRA | **7.16 GB** | OOM |
| Gemma-4-26B-A4B QLoRA | **8.47 GB** | OOM |
| OLMoE-1B-7B QLoRA | 2.57 GB (from 5.97) | 5.97 GB |

Cost is ~+11 % s/step (OLMoE; pinned-memory H2D copies overlap compute for the larger models).
Correctness: the frozen 4-bit forward is bit-identical with offload on/off, including through
`use_reentrant=False` gradient-checkpoint recompute (the recompute re-fetches the same quantized
bytes; adapters stay resident and trainable).

## Proposed shape

An expert-granularity mode that composes with `quantize_moe_experts` — conceptually
`layer_offloading` restricted to the expert parametrizations: pin the packed 4-bit expert storage
in CPU RAM at load, keep one layer's experts on GPU, prefetch the next layer's on a side stream
during compute, evict after use. Scope for a first PR: **single-GPU only** (explicitly out of scope:
FSDP/DDP interaction, expert-parallel), gated behind a config flag, with the
checkpointing-recompute path covered by tests.

I'm happy to write the PR against axolotl's parametrization representation (either in-tree as an
extension of the layer_offloading mixin, or as an `axolotl.integrations` plugin — maintainer
preference). The numbers above are reproducible from the linked repo; the same primitive is also
proposed upstream in bitsandbytes (bitsandbytes-foundation/bitsandbytes#1965, draft until their
v0.50), and an in-tree implementation here could switch to that once it lands.

**Comment by NanoCode012** [comment]
If you can provide a PR to axolotl showing this, we can take a look at it. Could you also link the wandb charts + perf tables for this with more details? We've been making a lot of changes to the offloading. Your linked repo looks like just a mini bench atm?

**Comment by pjordanandrsn** [comment]
Thanks — fair on both counts. I re-ran it as a clean, matched A/B on a real card so it's not "just a mini bench": single **12 GB RTX A2000**, OLMoE-1B-7B QLoRA, r=8, seq 256, 60 steps, **same seed + data order for both arms**, per-step metrics to wandb + JSONL.

### VRAM — the point of the feature

| config | loaded GPU | peak GPU | held-out eval (before → after) |
|---|--:|--:|--:|
| experts resident on GPU | 4.70 GB | 6.00 GB | 1.645 → 1.221 |
| experts offloaded | **1.08 GB** | **2.60 GB** | 1.645 → 1.227 |

Peak **−57 %** (6.00 → 2.60 GB), load-time footprint **−77 %** — the 4-bit experts stream from pinned CPU RAM with one layer resident. OLMoE already fits on the card, so this is the measurable proxy; it's the *same* mechanism that lets the 30B-A3B / 26B-A4B configs from the issue fit at all on 12 GB.

![vram](https://raw.githubusercontent.com/pjordanandrsn/experts4bit-qlora/main/ab-telemetry/charts/vram.png)

### Convergence is preserved

Both arms share a seed, so BEFORE is identical (1.6448) and the training-loss curves track each other step-for-step, converging to ~1.22 either way (Δ **−0.4235** off vs **−0.4178** on — the small gap is fp noise in the dequant/reload path, incl. the `use_reentrant=False` gradient-checkpoint recompute, not divergence).

![loss](https://raw.githubusercontent.com/pjordanandrsn/experts4bit-qlora/main/ab-telemetry/charts/loss_curve.png)

### Throughput

Median **+30 % s/step** with offload on *on this box* — but treat that as an upper bound: it was a shared host under heavy CPU contention (2-core cgroup cap, host load ~55 on 12 cores), and the offload cost is the per-layer H2D copy, which can't overlap compute when the CPU is starved. Uncontended I've measured ~+11 % (repo README). The step-time chart shows offload-on riding a roughly constant ~8–12 s above off — a memory-for-compute trade, as expected, not a regression.

![throughput](https://raw.githubusercontent.com/pjordanandrsn/experts4bit-qlora/main/ab-telemetry/charts/throughput.png)

All artifacts — per-step JSONL, the two `wandb sync`-able offline runs, the chart generator, the A/B runner + methodology — are in [`ab-telemetry/`](https://github.com/pjordanandrsn/experts4bit-qlora/tree/main/ab-telemetry); the instrumented trainer is branch [`telemetry-wandb-jsonl`](https://github.com/pjordanandrsn/experts4bit-qlora/tree/telemetry-wandb-jsonl).

### PR

Opened: **#3797**. I went with a standalone `axolotl.integrations.expert_offload` plugin (self-contained, mirrors the `expert_parallel` integration) — happy to refold it into the `layer_offloading` mixin instead if you'd prefer that shape. Single-GPU only (refuses under FSDP / DeepSpeed / expert-parallel), gated behind `expert_offload: true`, with the gradient-checkpoint recompute path under test on CPU and against real bitsandbytes `Linear4bit` on CUDA (23/23 on the A2000). It sits on current `main` and merges clean.

**Comment by NanoCode012** [comment]
Thanks for the PR. I like the idea, but I would like to see benchmarks using axolotl, using your PR preferably. I checked the branch and can only see your training loop. The wandb logs is not from our trainer.

**Comment by rehan243** [comment]
Oh yeah, we've hit this exact wall trying to squeeze 30B+ MoEs onto 16 GB A5000s. The GPU VRAM just vanishes into the expert layers, especially with fused 4-bit. We worked around it by hacking together a super-janky custom prefetch pipeline, but it wasn’t pretty. Your approach looks waaay cleaner.

For us, chunking the expert weights into pinned CPU buffers and overlapping the H2D with compute made a huge difference. Here’s a snippet of the prefetch logic we used (pseudo-ish, but you get it):

```python
def prefetch_experts(next_expert_weights, stream):
    with torch.cuda.stream(stream):
        next_expert_weights.copy_()
```

We’d call this during the forward pass of expert N, loading N+1's weights into pinned memory. Got the overhead down to ~9% on a 30B MoE with a similar setup. Curious, are you using PyTorch’s `torch.cuda.Stream` for your prefetch, or did you roll something custom? And does your implementation handle multi-GPU setups cleanly? We ran into some weird NCCL contention issues with the H2D overlap.

**Comment by pjordanandrsn** [comment]
Good questions—both.

**(a) Prefetch: plain `torch.cuda.Stream`, nothing custom.** One side stream per device; the H2D copy is issued on the side stream, an event is recorded, and the compute stream waits on that event before touching the expert. Copied tensors get `record_stream` so the allocator doesn't recycle them under the copy. The nuance worth knowing: the *training* path in the PR is deliberately synchronous single-slot—stage N evicts N−1 on the current stream. With `gradient_checkpointing` + `use_reentrant: False`, the backward recompute re-stages every block anyway, so a double-buffered prefetch buys nothing there—that overlap path is the inference side of the source repo. Your ~9% with overlap lines up with the ~11% we measured uncontended on PCIe 3.0 for that path; trainer-native training-path overhead at seq 2048 is bigger (~35%, table in the PR thread) because each stage is on the critical path by design.

**(b) Multi-GPU: plain DDP is now supported and measured.** How it composes:

- Each rank keeps its own pinned CPU home for the offloaded experts—host RAM scales ×`world_size`, stated in the docs, not hidden. OLMoE homes 3.22 GB/rank.
- Offloaded expert weights are registered on DDP's parameter ignore list before DDP construction, so the initial module-state sync never touches the evicted placeholders and the frozen weights never enter gradient buckets.
- The only collective traffic is the LoRA adapter all-reduce—tens of MB per step—so NCCL-vs-H2D contention was the open question. Measured (2×RTX 3090, one pod, seed-matched): offload costs +35.5% s/it single-GPU and +34.1% per-rank under DDP. The delta is ~−1 point—no measurable contention at this scale; the all-reduce is too small to fight the staging traffic.
- FSDP / DeepSpeed / expert-parallel stay refused by the validator—those shard or move the same weights this plugin evicts; composing them is a different design.

Full four-arm table (single/DDP × resident/offload, all through `axolotl train` on the branch) is in the PR thread; configs and raw traces: https://github.com/pjordanandrsn/experts4bit-qlora/tree/main/ab-telemetry/axolotl-ab

No rush on any of this—happy to go deeper on the stream mechanics if useful.

**Comment by rehan243** [comment]
Ah, interesting — so you’re sticking to the default stream sync for prefetch, got it. That makes sense given your setup. Yeah, that overlap on PCIe 3.0 sounds pretty close to what I saw, so I’d guess the gains would scale a lot better on higher bandwidth links. Did you try any async techniques beyond just the side stream?

**Comment by pjordanandrsn** [comment]
The axolotl-native benchmarks requested are now on #3797 (four-arm `axolotl train` table: single/DDP × resident/offload — loaded memory −76%, +35% s/it). Anything else needed to clear `waiting for reporter` here and the `hold` on #3797?

**Comment by rehan243** [comment]
Hm, that's a solid improvement. Yeah, that version issue bit us too. I'd check if the latest axolotl build has any fixes for the offload path.

If you're still seeing issues, there's a PR open for some more aggressive memory optimizations. Might be worth a look. Lmk if that helps, or if you need more details on our prefetch hack.

**Comment by NanoCode012** [comment]
@pjordanandrsn , sorry for late update. We've been working to get a release out so a bit limited on resources to check on this. We'll get back to you when possible. Thanks for running the e2e axolotl train

**Comment by pjordanandrsn** [comment]
@rehan243, beyond the side stream: no. I tested speculative prefetch and the router isn’t predictable enough for it to pay, so the trainer path stays synchronous single-slot (mechanics in my comment above).

The ~9% is what I’d want to understand: which link (disk→pinned? pinned→GPU?), train or decode, how you pick what to stage, and how was it measured? Runnable branch plus numbers and I’ll dig in.

**Comment by rehan243** [comment]
No worries, totally get the release grind eating up time. Just ping me if you want me to test anything specific or dig into the prefetch logic deeper once you have bandwidth. Also, if you're curious, I found some neat perf tweaks in the latest PyTorch nightly that helped with async transfers on CPU offloading, might be worth a peek. Lmk if that helps or if you want me to share the snippet.
